app/models/journals/scopes/with_sequence_version.rb

```diff
@@ -0,0 +1,55 @@
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2010-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+
+module Journals::Scopes
+  module WithSequenceVersion
+    extend ActiveSupport::Concern
+
+    class_methods do
+      def with_sequence_version
+        joins(
+          <<~SQL.squish
+            JOIN LATERAL (
+              SELECT
+                journals_rank.id,
+                ROW_NUMBER() OVER (PARTITION BY journable_id ORDER BY version ASC) sequence_version
```

Comment:
✨ Amazing.

---

One thing to note is `row_number` [_"generates a new row number for every row, regardless of duplicates within a partition."_](https://blog.jooq.org/sql-trick-row_number-is-to-select-what-dense_rank-is-to-select-distinct/) so we can only rely on the `sequence_version` when journables are already scoped- otherwise we'll get duplicates.

Reply:
Good pointer. In this case, I think we are safe given that the subselect is filtered down to the same journable (id and type) by the `where`s